File: Data Structures & Algorithms/implement-prefix-tree/submission-1.go
type TrieNode struct {
	children [26]*TrieNode
	isEnd    bool
}

type PrefixTree struct {
	root *TrieNode
}

func Constructor() PrefixTree {
	return PrefixTree{
		root: &TrieNode{},
	}
}

func (this *PrefixTree) Insert(word string) {
	node := this.root

	for i := 0; i < len(word); i++ {
		index := word[i] - 'a'

		if node.children[index] == nil {
			node.children[index] = &TrieNode{}
		}

		node = node.children[index]
	}

	node.isEnd = true
}

func (this *PrefixTree) Search(word string) bool {
	node := this.root

	for i := 0; i < len(word); i++ {
		index := word[i] - 'a'

		if node.children[index] == nil {
			return false
		}

		node = node.children[index]
	}

	return node.isEnd
}

func (this *PrefixTree) StartsWith(prefix string) bool {
	node := this.root

	for i := 0; i < len(prefix); i++ {
		index := prefix[i] - 'a'

		if node.children[index] == nil {
			return false
		}

		node = node.children[index]
	}

	return true
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-1.java
class WordDictionary {

    class TrieNode {
        TrieNode[] children = new TrieNode[26];
        boolean isEnd = false;
    }

    private TrieNode root;

    public WordDictionary() {
        root = new TrieNode();
    }

    public void addWord(String word) {
                TrieNode node = root;

        for (char c : word.toCharArray()) {
            int index = c - 'a';

            if (node.children[index] == null) {
                node.children[index] = new TrieNode();
            }

            node = node.children[index];
        }

        node.isEnd = true;
    }

    public boolean search(String word) {
        return dfs(word, 0, root);
    }

    private boolean dfs(String word, int index, TrieNode node) {
        if (node == null) return false;
        if (index == word.length()) return node.isEnd;

        char c = word.charAt(index);

        if (c == '.') {
            // Wildcard: thử TẤT CẢ 26 nhánh con
            for (TrieNode child : node.children) {
                if (dfs(word, index + 1, child)) {
                    return true; // chỉ cần 1 nhánh đúng là đủ
                }
            }
            return false; // không nhánh nào khớp
        } else {
            // Ký tự bình thường: chỉ đi đúng 1 nhánh
            return dfs(word, index + 1, node.children[c - 'a']);
        }
    }
}
